radix-engine/src/model/vault.rs

```diff
@@ -56,39 +49,44 @@ impl Vault {
         &mut self,
         ids: &BTreeSet<NonFungibleId>,
     ) -> Result<Bucket, VaultError> {
-        let this_container = self.borrow_container()?;
-
         Ok(Bucket::new(
-            this_container
+            self.borrow_container_mut()
                 .take_non_fungibles(ids)
                 .map_err(VaultError::ResourceContainerError)?,
         ))
     }
 
-    pub fn liquid_amount(&self) -> ResourceAmount {
-        self.container.liquid_amount()
-    }
-
     pub fn resource_def_id(&self) -> ResourceDefId {
-        self.container.resource_def_id()
+        self.borrow_container().resource_def_id()
     }
 
     pub fn resource_type(&self) -> ResourceType {
-        self.container.resource_type()
+        self.borrow_container().resource_type()
     }
 
-    /// Creates another `Rc<ResourceContainer>` to the container
-    pub fn reference_container(&self) -> Rc<ResourceContainer> {
-        self.container.clone()
+    pub fn total_amount(&self) -> Decimal {
+        self.borrow_container().total_amount()
+    }
+
+    pub fn total_ids(&self) -> Result<BTreeSet<NonFungibleId>, VaultError> {
+        self.borrow_container()
+            .total_ids()
+            .map_err(VaultError::ResourceContainerError)
+    }
+
+    pub fn is_locked(&self) -> bool {
```

Comment:
Thoughts on making these methods implemented in a default Trait?

Reply:
You mean a trait for these getters that forward request to the "resource container"?

I'm not sure how this is possible if I want to keep composition, i.e. bucket/vault/worktop contains one or more resource containers (implementation). This essentially is a good example of the "delegate call" idea.

An alternative approach is remove all these getters and redirect clients to use the `borrow_container` directly.

What do you think?

Reply:
More of the latter since a whole subset of functionality is essentially the same. But, I don't think it's too big of a deal for now.